week-4/E-arrays-of-objects/exercise-2.js

```diff
@@ -7,50 +7,83 @@ Each destination has a name, a distance from Glasgow, and a list of transportati
 3) Print in the console all the destination names more than 300 kms far away and reachable by train.
 */
 
-
 var destination1 = {
-    destinationName: "Edinburgh",
-    distanceKms: 80,
-    transportations: ["car", "bus", "train"]
+  destinationName: "Edinburgh",
+  distanceKms: 80,
+  transportations: ["car", "bus", "train"]
 };
 
 var destination2 = {
-    destinationName: "London",
-    distanceKms: 650,
-    transportations: ["car", "bus", "train"]
+  destinationName: "London",
+  distanceKms: 650,
+  transportations: ["car", "bus", "train"]
 };
 
 var destination3 = {
-    destinationName: "Paris",
-    distanceKms: 900,
-    transportations: ["train", "plane"]
+  destinationName: "Paris",
+  distanceKms: 900,
+  transportations: ["train", "plane"]
 };
 
 var destination4 = {
-    destinationName: "Dublin",
-    distanceKms: 350,
-    transportations: ["plane", "ferry"]
+  destinationName: "Dublin",
+  distanceKms: 350,
+  transportations: ["plane", "ferry"]
 };
 
-var travelDestinations = [destination1, destination2, destination3, destination4];
-
-/* 
-DO NOT EDIT ANYTHING ABOVE THIS LINE
-WRITE YOUR CODE BELOW
-*/
+var travelDestinations = [
+  destination1,
+  destination2,
+  destination3,
+  destination4
+];
+//I tried to create new Array Method which covers both .filter() and .map() methods.
 
+Array.prototype.filterPlus = function(callback, propertyName) {
+  var output = [];
+  for (var element of this) {
+    if (callback(element)) {
+      output.push(element[propertyName]);
+    }
+  }
+  return output;
+};
 
-var destinationNamesWithin500Kms = // Complete here
-
-var destinationNameReachableByFerry = // Complete here
-
-var destinationNamesMoreThan300KmsAwayByTrain = // Complete here (PRINT THE RESULT IN THE CONSOLE USING FOREACH)
+var destinationNamesWithin500Kms = travelDestinations.filterPlus(
+  inside500,
+  "destinationName"
+);
+function inside500(destination) {
+  return destination.distanceKms < 500;
+} // Complete here
 
+var destinationNameReachableByFerry = travelDestinations.find(destination =>
+  destination.transportations.includes("ferry")
+).destinationName; // Complete here
 
```

Comment:
This does work when only one of the destinations is reachable by ferry. However, if our list was larger and multiple destinations were reachable by ferry, it would still only pass me one value. You may have to consider an additional array method for this to resolve this problem.

Reply:
As the instruction for this part `2) Find a destination name reachable by ferry.` says to find only one destination name I used `find()` method otherwise I would use `filter()+map()` methods or `filterPlus()` method.